### ISSUE TITLE
Is this anomaly detection approach doable?

### ISSUE THREAD

**Issue by jeff-brooks** [dotnet/machinelearning#4843]
First time creating an issue, apologies in advance if I'm doing it wrong.

Anomaly detection in general seems to go against my expectations. I've looked at every example I can find for anomaly detection and they all follow the pattern of having some dataset that the anomaly detector examines to see what anomalies it can find within that data set.

My needs are different than that. I want to provide the anomaly detector a set of data that has no anomalies. This dataset should be used to train and create a model for comparison against a second data sets to see what anomalies it detects in this second data set. 

Thus far, my attempts have ended up with an anomaly detector that finds "anomalies" in my first data set, which is not valid in my scenario. The first data set should be considered all valid (non-anomaly) values, no matter what seemingly spikes or change points may exist within it. The second data set should be examined for spikes or change points through the lens of was seen in the first data set.

Can this be done? I've been trying to make it work on my own but it seems like either this is not how anomaly detection was intended to work, or I'm completely off base in my implementation.

**Comment by frank-dong-ms** [comment]
Thanks for reaching out.

What you described is not typical anomaly detection scenario as far as I know.
Anomaly detection can be done in various approaches but basic idea of anomaly detection is based on some basic concepts: data trend, data period(auto detect or user assign), tolerance to randomness thus to determine one data point is anomaly or not based on historical data pattern.

Maybe you should consider use other approach to achieve your purpose. If you want please describe you scenario in detail and we can see what we can offer from ML.NET, thanks.

**Comment by jeff-brooks** [comment]
@frank-dong-ms , Thanks for that information.

The essence of my purpose is captured in my initial post, but I can add some more context here. I have a website which tracks inventory of stock in a physical warehouse. As the amount of stock for an item changes, users will go into the system and manually update how much stock is in the warehouse for a given item. This system is several years old and it has both a transactional database which keeps track of current inventory counts and a historical database that tracks of how much stock was available within a historical timeframe.

I'm looking into the ML.net library to see if it can be used to do some input validation when a user goes into the system and enters a new value for a stock item. Basically, checking for potential fat-finger mistakes. Say an item typically ranges between 0 and 100. That it has been within this range for several years, and then all of a sudden, a user tries to enter a value way out of the norm, say 100,000. That would be something I would like to check for and build some robustness around. 

My thought was that maybe I could pull from the historical database all of the previous counts for a particular item that have been entered. These values should be considered non-anomalies despite any seeming spikes because it's historical data and I'm not intersted in finding potential fat-fingered entries in historical data. I would like to train a model with this historical data so that machine learning develops a picture of what the normal range of values looks like, and then use that information to predict whether or not a newly entered value is outside the norm or not. So, in this scenario, the question I would be trying to ask of ML.net would be, "Is this abnormal input? Is this a potential fat-fingered input value?"

So, in this scenario, I have no interest in finding spikes in my historical data. Because that doesn't help with doing input validation. I'm looking to find a way to either prompt the user to take a second look at their entry because it "looks odd", or maybe prompt a second set of eyes from a different person to double-check.

I hope this helps to convey my use case?

**Comment by yaeldekel** [comment]
Hi @jeff-brooks . The scenario that @frank-dong-ms described is a specific kind of anomaly detection, called time series anomaly detection. Your scenario is not time series based, but it still seems like an anomaly detection scenario to me. Anomaly detection trainers of the kind that you are looking for, train a model on data that is all non-anomalies, and can be used to predict anomalies on new data. However, the way it works is that the model computes a score for each example, and you can decide the threshold to determine which score you consider to be an anomaly, based on your tolerance for false positives/negatives. Given the threshold, the model might make some errors on the training data, but that is to be expected.

ML.NET has an anomaly detection estimator, called [randomized PCA](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.PCA/PCACatalog.cs#L67-L78) that you can try out to see if it fits your use case.

**Comment by jeff-brooks** [comment]
Hi @yaeldekel , thanks for that info.

I have taken a look at the Randomized PCA trainer. This was the most hopeful example I found: https://github.com/dotnet/machinelearning-samples/commit/7cb26f32ae412e015c2a3e2d8bccf97a64055a5f#diff-db5b69ae0ce54e54d388e93a3a1b9f48

I followed it but unfortunately didn't notice much difference between the Randomized PCA trainer and the time series anomaly detector transforms when trying to predict anomaly detection based off of some dataset of non-anomaly values.

**Comment by najeeb-kazmi** [comment]
@jeff-brooks Randomized PCA approximates the subspace for the normal class using the top `RandomizedPcaTrainer.Options.Rank` eigenvectors of your training data. The score for each example is calculated as the difference between the L2-norm of the initial feature vector and the L2-norm of the feature vector projected onto this subspace. A score closer to 0 would indicate the feature vector and its projection are not very different (i.e. not an anomaly).

As Yael commented, this is to be expected on the training set purely because some training examples are going to have feature vectors somewhat different than their projections. This is fine. Once a model is trained on the normal examples in the training set, you can use it to calculate the anomaly score on the test set. You can then choose your own threshold on the score to define an anomalous example (depending on your tolerance for false positives/negatives). Choosing an appropriate rank should help with some training examples being labeled as anomalies.

See also [these examples](https://github.com/dotnet/machinelearning/tree/master/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection).

Please reopen if something remains unanswered.

**Comment by jeff-brooks** [comment]
Thanks @najeeb-kazmi . I'll take another look at the rank configuration in my prototype.

**Comment by SphoortiMetri** [comment]
Hi @jeff-brooks 

I am trying to write an application of anomaly detection for a use case that is conceptually similar to yours. I did not have any luck with randomized PCA. Do you have any additional inputs? Thanks in advance!
